pkg/controller/mizar/node_controller.go

```diff
@@ -0,0 +1,307 @@
+			oldResource := oldObject.(*v1.Node)
+			newResource := newObject.(*v1.Node)
+			oldKey := c.genKey(oldResource)
+			newKey := c.genKey(newResource)
+			if oldKey == newKey {
+				klog.Infof(c.logEventMessage("No actual change in nodes, discarding: ", newKey))
+			} else {
+				command, err := c.compareKey(oldKey, newKey)
+				if err != nil {
+					klog.Errorf(c.logInfoMessage("Unexpected string in queue; discarding: " + newKey))
+				} else {
+					c.Enqueue(command, newKey)
+				}
+				fmt.Printf("NODE_UPDATE: %s, %s", command, newKey)
+			}
+		},
+		DeleteFunc: func(object interface{}) {
+			resource := object.(*v1.Node)
+			key := c.genKey(resource)
+			c.Enqueue(NODE_DELETE, key)
+			fmt.Printf("NODE_DELETE: %s", key)
+		},
+	})
+	return c, nil
+}
+
+// Run starts an asynchronous loop that detects events of cluster nodes.
+func (c *NodeController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+	klog.Infof(c.logInfoMessage("Starting node controller"))
+	klog.Infoln(c.logInfoMessage("Waiting cache to be synced"))
+	if ok := cache.WaitForCacheSync(stopCh, c.informerSynced); !ok {
+		klog.Fatalln(c.logInfoMessage("Timeout expired during waiting for caches to sync."))
+	}
+	klog.Infoln(c.logInfoMessage("Starting workers..."))
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+	<-stopCh
+	klog.Info(c.logInfoMessage("Shutting down node controller"))
+}
+
+// Enqueue puts key of the node object in the work queue
+func (c *NodeController) Enqueue(command string, key string) {
+	commandKey := c.genCommandKey(command, key)
+	c.queue.Add(commandKey)
+}
+
+// Dequeue an item and process it
+func (c *NodeController) runWorker() {
+	for {
+		item, queueIsEmpty := c.queue.Get()
+		if queueIsEmpty {
+			break
+		}
+		c.process(item)
+	}
+}
+
+// Parsing a item key and call gRPC request
+func (c *NodeController) process(item interface{}) {
+	defer c.queue.Done(item)
+	commandKey, ok := item.(string)
+	if !ok {
+		klog.Errorf(c.logInfoMessage("unexpected item in queue: " + commandKey))
+		c.queue.Forget(item)
+		return
+	}
+	command, key, err := c.parseCommandKey(commandKey)
+	if err != nil {
+		klog.Errorf(c.logInfoMessage("Unexpected item in queue: " + commandKey))
+		c.queue.Forget(item)
+		return
+	}
+	nodeKind, nodeTenant, nodeName, nodeStatusType, nodeStatus, nodeAddressType, nodeAddress, err := c.parseKey(key)
+	if err != nil {
+		klog.Errorf(c.logInfoMessage("Unexpected string in queue; discarding: ") + key)
+		c.queue.Forget(item)
+		return
+	}
+	if nodeKind != NODE_KIND {
+		klog.Errorf(c.logInfoMessage("Unexpected object in queue; discarding: ") + key)
+		c.queue.Forget(item)
+		return
+	}
+	klog.V(5).Infof(c.logInfoMessage("Processing a node: ")+"%s/%s/%s/%s/%s/%s/%s", nodeKind, nodeTenant, nodeName, nodeStatusType, nodeStatus, nodeAddressType, nodeAddress)
+	c.gRPCRequest(command, nodeAddress)
+	c.queue.Forget(item)
```

Comment:
processing the item might fail. When it fails, usually the item should put back in queue to let controller has chance to do it again. Please consider make the code change, to differ whether processing was successful or not, and hence to remove the item out of the queue, or re-add it to.

Reply:
Thank you for the comment. This is critical. If process is failed, it will add the key to queue. Here is the change
result, err := c.gRPCRequest(eventType, nodeName, nodeAddress)
	if !result {
		klog.Errorf(c.logMessage("Failed a node processing: ", key))
		c.queue.Add(item)
	} else {
		klog.Infof(c.logMessage(" Processed a node: ", key))
		c.queue.Forget(item)
	}